src/v/serde/serde.h

```diff
@@ -0,0 +1,260 @@
+                                                                               \
+    template<typename T>                                                       \
+    struct SERDE_TOKEN_PASTE(                                                  \
+      help_has_,                                                               \
+      name)<T, std::void_t<decltype(std::declval<T>().name(__VA_ARGS__))>>     \
+      : std::true_type {};                                                     \
+                                                                               \
+    template<typename T>                                                       \
+    inline constexpr auto const SERDE_TOKEN_PASTE(                             \
+      has_, name) = SERDE_TOKEN_PASTE(help_has_, name)<T>::value;
+
+static auto iob = iobuf{};
+static auto iop = iobuf_parser{std::move(iob)};
+
+HAS_MEMBER_FN(serde_read, iop, version_t{0}, version_t{0}, size_t{0U})
+HAS_MEMBER_FN(serde_write, iob)
+HAS_MEMBER_FN(serde_async_read, iop, version_t{0}, version_t{0}, size_t{0U})
+HAS_MEMBER_FN(serde_async_write, iob)
+
+template<typename T>
+inline constexpr auto const is_serde_compatible_v =
+  is_envelope_v<T> || std::is_scalar_v<T> || std::is_enum_v<T> ||
+  reflection::is_std_vector_v<T> || reflection::is_named_type_v<T> ||
+  reflection::is_ss_bool_v<T> || std::is_same_v<T, std::chrono::milliseconds> ||
+  std::is_same_v<T, iobuf> || std::is_same_v<T, ss::sstring> ||
+  reflection::is_std_optional_v<T>;
+
+using header_t = std::tuple<version_t, version_t, size_t>;
+
+void write_envelope() {}
+
+template<typename T>
+void write(iobuf& out, T const& t) {
+    using Type = std::decay_t<T>;
+    static_assert(has_serde_write<Type> || is_serde_compatible_v<Type>);
+
+    if constexpr (is_envelope_v<Type>) {
+        write(out, Type::__version);
+        write(out, Type::__compat_version);
+
+        auto size_placeholder = out.reserve(
+          sizeof(typename Type::envelope_size_t));
+        auto const size_before = out.size_bytes();
+
+        if constexpr (has_serde_write<Type>) {
+            t.serde_write(out);
+        } else {
+            envelope_for_each_field(t, [&out](auto& f) { write(out, f); });
+        }
+
+        auto const written_size = out.size_bytes() - size_before;
+        if (
+          written_size >
+          std::numeric_limits<typename Type::envelope_size_t>::max()) {
+            throw std::system_error{
+              make_error_code(error_codes::envelope_too_big)};
+        }
+        auto const size = ss::cpu_to_le(
+          static_cast<typename Type::envelope_size_t>(written_size));
+        size_placeholder.write(
+          reinterpret_cast<char const*>(&size),
+          sizeof(typename Type::envelope_size_t));
+    } else if constexpr (std::is_enum_v<Type>) {
+        write(out, static_cast<std::underlying_type_t<Type>>(t));
+    } else if constexpr (std::is_scalar_v<Type>) {
```

Comment:
I think that `is_scalar_v` will also handle `bool`? in that case, it seems like can't necessarily rely on `bool` having a predictable representation? in this case, it seems like it is going to be dependent which particular overload of `auto le_t = ss::cpu_to_le(t);`is chosen?

Reply:
related to is_enum case above because is_scalar is true for Type is enum

Reply:
ss::cpu_to_le(t); may not give predictable behavior in futures version of seastar when given a bool